src/frontend/src/utils/condition.rs

```diff
@@ -292,45 +301,43 @@ impl Condition {
                 if let Some((input_ref, const_expr)) = expr.as_eq_const() &&
                     let Ok(const_expr) = const_expr.cast_implicit(input_ref.data_type) {
                     assert_eq!(input_ref.index, order_column_ids[i]);
-                    let Ok(Some(value)) = const_expr.eval_row_const() else {
-                        // column = NULL or failed to eval
-                        return always_false();
+                    let Some(value) = const_expr.eval_row_const()? else {
+                        // column = NULL
+                        return Ok(false_cond());
                     };
                     if !eq_conds.is_empty() && eq_conds.into_iter().all(|l| l != value) {
-                        return always_false();
+                        return Ok(false_cond());
                     }
                     eq_conds = vec![value];
                 } else if let Some((input_ref, in_const_list)) = expr.as_in_const_list() {
                     assert_eq!(input_ref.index, order_column_ids[i]);
                     let mut scalars = HashSet::new();
                     for const_expr in in_const_list {
-                        // The cast should succeed, because otherwise the input_ref is casted
-                        // and thus `as_in_const_list` returns None.
```

Comment:
Certain parts of the comment still applies to the safety of `unwrap` after `cast_implicit`, because a casted input_ref on the left cannot enter this path, meaning the left side is guaranteed to have a less restrictive type than values in const list.

Reply:
Oh yes, it explains the `unwrap` after `cast_implicit`, not after `eval_row_const`. I forgot it. 🥸